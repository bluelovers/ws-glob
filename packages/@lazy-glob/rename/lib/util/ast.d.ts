import { IPartAstInterface, IQuestionMark, IAsterisk } from '../types';
export declare function isAsterisk(g: any | IPartAstInterface): g is IAsterisk;
export declare function isQuestionMark(g: any | IPartAstInterface): g is IQuestionMark;
