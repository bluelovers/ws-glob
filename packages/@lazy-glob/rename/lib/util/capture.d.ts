export declare function capture(names: string[], glob: string): import("./globGroupsCollection").globGroupsCollectionInterface[];
