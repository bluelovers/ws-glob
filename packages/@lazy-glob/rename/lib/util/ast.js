"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isQuestionMark = exports.isAsterisk = void 0;
const types_1 = require("../types");
function isAsterisk(g) {
    return (g.type === 'wildcard' && g.pattern === types_1.EnumWildcard.Asterisk);
}
exports.isAsterisk = isAsterisk;
function isQuestionMark(g) {
    return (g.type === 'wildcard' && g.pattern === types_1.EnumWildcard.QuestionMark);
}
exports.isQuestionMark = isQuestionMark;
//# sourceMappingURL=ast.js.map