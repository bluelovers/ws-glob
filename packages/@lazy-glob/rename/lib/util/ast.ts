import { IPartAstInterface, IQuestionMark, EnumWildcard, IAsterisk } from '../types';

export function isAsterisk(g: any | IPartAstInterface): g is IAsterisk
{
	return (g.type === 'wildcard' && g.pattern === EnumWildcard.Asterisk);
}

export function isQuestionMark(g: any | IPartAstInterface): g is IQuestionMark
{
	return (g.type === 'wildcard' && g.pattern === EnumWildcard.QuestionMark);
}
