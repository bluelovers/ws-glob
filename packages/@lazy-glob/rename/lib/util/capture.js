"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * For each string in names array, return a globGroupsCollection (an object implementing globGroupsCollectionInterface) containing the substring matched by each part of the glob pattern.
 * Example of parts of a glob: "?omer.*" => Parts: ["?"] ["omer."] ["*"]
 * @method capture
 * @param names {String[]} List of names
 * @param glob {String} The glob pattern to be applied to names
 * @return {globGroupsCollection[] | null} An array of globGroupsCollection objects. Returns null if invalid parameters or names list is empty.
 */
const globGroupsCollection_1 = require("./globGroupsCollection");
function capture(names, glob) {
    if (!Array.isArray(names) || !names.length) {
        return null;
    }
    if ((typeof glob !== 'string') || !glob.length) {
        return null;
    }
    // For each name received, create an object of type 'globGroupsCollection' containing the match for each part of the glob
    return names.map(function buildGlobGroupsCollection(name) {
        let groupsObj = globGroupsCollection_1.globGroupsCollectionFactory();
        groupsObj.initGroups(glob);
        groupsObj.buildGroups(name);
        return groupsObj;
    });
}
//# sourceMappingURL=capture.js.map