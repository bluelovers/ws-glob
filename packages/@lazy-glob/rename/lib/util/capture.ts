/**
 * For each string in names array, return a globGroupsCollection (an object implementing globGroupsCollectionInterface) containing the substring matched by each part of the glob pattern.
 * Example of parts of a glob: "?omer.*" => Parts: ["?"] ["omer."] ["*"]
 * @method capture
 * @param names {String[]} List of names
 * @param glob {String} The glob pattern to be applied to names
 * @return {globGroupsCollection[] | null} An array of globGroupsCollection objects. Returns null if invalid parameters or names list is empty.
 */
import { globGroupsCollectionFactory } from './globGroupsCollection';

function capture(names: string[], glob: string)
{
	if (!Array.isArray(names) || !names.length)
	{
		return null;
	}

	if ((typeof glob !== 'string') || !glob.length)
	{
		return null;
	}

	// For each name received, create an object of type 'globGroupsCollection' containing the match for each part of the glob
	return names.map(function buildGlobGroupsCollection(name)
	{
		let groupsObj = globGroupsCollectionFactory();
		groupsObj.initGroups(glob);
		groupsObj.buildGroups(name);
		return groupsObj;
	});
}
