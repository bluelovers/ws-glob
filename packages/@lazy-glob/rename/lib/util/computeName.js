"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeName = void 0;
const deconstruct_1 = require("./deconstruct");
const capture_1 = require("./capture");
const countWildcards_1 = require("./countWildcards");
const types_1 = require("../types");
/**
 * Builds new names based on provided glob patterns.
 * @param filenames {String[] | string} List of original names, or a single name (string), to compute new names for
 * @param srcGlob {String} Glob pattern used to match the original names
 * @param dstGlob {String} Glob pattern used to contruct new names
 * @return {String[]} List of new names. String is empty if new name could not be computed.
 * @throws {Error} An Error object
 */
function computeName(filenames, srcGlob, dstGlob) {
    if ((typeof srcGlob !== 'string' || !srcGlob) || (typeof dstGlob !== 'string' || !dstGlob)) {
        throw new TypeError('Invalid type for glob pattern! Must be a non-empty string.');
    }
    if (typeof filenames === 'string') {
        filenames = [filenames];
    }
    else {
        if (Array.isArray(filenames)) {
            if (!filenames.length) {
                throw new TypeError('Invalid names list. Array must be non-empty.');
            }
        } // names is not an array, nor a string
        else {
            throw new TypeError('Invalid type for names! Must be a string or an Array of string.');
        }
    }
    // Deconstruct the source glob into literal and wildcard parts
    let srcGlobParts = (0, deconstruct_1.deconstruct)(srcGlob, { collapse: true });
    if (!srcGlobParts.length) {
        throw new Error('Unexpected error while parsing glob.');
    }
    // Deconstruct the destination glob into literal and wildcard parts
    let dstGlobParts = (0, deconstruct_1.deconstruct)(dstGlob, { collapse: false });
    if (!dstGlobParts.length) {
        throw new Error('Unexpected error while parsing glob.');
    }
    let srcWildcardsCount = (0, countWildcards_1.countWildcards)(srcGlobParts);
    let dstWildcardsCount = (0, countWildcards_1.countWildcards)(dstGlobParts);
    // Check if destination glob has more * and/or ? wildcards than source glob does...
    if ((dstWildcardsCount.stars > srcWildcardsCount.stars) || (dstWildcardsCount.questions > srcWildcardsCount.questions)) {
        throw new Error('Invalid glob pattern. Destination glob contains incorrect number or type of wildcard. (** is treated as * in source glob.)');
    }
    // extract matches for each wildcard (and literal) parts of the source glob
    let srcCaptureGroupsArray = (0, capture_1.capture)(filenames, srcGlob);
    if (!srcCaptureGroupsArray) {
        throw new Error('Unexpected error while extracting glob matches.');
    }
    let computedNames = filenames.map(function buildNewName(name, nameIdx) {
        // The source glob is required to match the original name in order to proceed building the new name
        if (!srcCaptureGroupsArray[nameIdx].hasMatch()) {
            return '';
        }
        let newName = '';
        let srcAsteriskList = srcCaptureGroupsArray[nameIdx].getAsterisk();
        let srcAsteriskIterator = (srcAsteriskList === null || srcAsteriskList === void 0 ? void 0 : srcAsteriskList.length) ? srcAsteriskList.entries() : null;
        let srcQuestionMarkList = srcCaptureGroupsArray[nameIdx].getQuestionMark();
        let srcQuestionMarkIterator = (srcQuestionMarkList === null || srcQuestionMarkList === void 0 ? void 0 : srcQuestionMarkList.length)
            ? srcQuestionMarkList.entries()
            : null;
        // Cycle through the parts of the destination glob to build the new name
        for (const destPart of dstGlobParts) {
            let srcGroup;
            switch (destPart) {
                case types_1.EnumWildcard.Asterisk:
                    srcGroup = srcAsteriskIterator.next();
                    // value[0] is the key, value[1] is the value, e.g. the actual captureGroup object.
                    // Append the * match to the new name construction
                    newName = newName + srcGroup.value[1].match;
                    break;
                case types_1.EnumWildcard.QuestionMark:
                    srcGroup = srcQuestionMarkIterator.next();
                    // value[0] is the key, value[1] is the value, e.g. the actual captureGroup object.
                    // Appends the ? match to the new name construction
                    newName = newName + srcGroup.value[1].match;
                    break;
                default: // For literal parts, simply append them to new name construction
                    newName = newName + destPart;
            }
        }
        return newName;
    });
    return computedNames;
}
exports.computeName = computeName;
//# sourceMappingURL=computeName.js.map