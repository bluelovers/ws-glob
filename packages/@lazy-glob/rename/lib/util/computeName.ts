import { deconstruct } from './deconstruct';
import { capture } from './capture';
import { countWildcards } from './countWildcards';
import { EnumWildcard, IPartsPattern } from '../types';

/**
 * Builds new names based on provided glob patterns.
 * @param filenames {String[] | string} List of original names, or a single name (string), to compute new names for
 * @param srcGlob {String} Glob pattern used to match the original names
 * @param dstGlob {String} Glob pattern used to contruct new names
 * @return {String[]} List of new names. String is empty if new name could not be computed.
 * @throws {Error} An Error object
 */
export function computeName(filenames: string | string[], srcGlob: string, dstGlob: string)
{
	if ((typeof srcGlob !== 'string' || !srcGlob) || (typeof dstGlob !== 'string' || !dstGlob))
	{
		throw new TypeError('Invalid type for glob pattern! Must be a non-empty string.');
	}
	if (typeof filenames === 'string')
	{
		filenames = [filenames];
	}
	else
	{
		if (Array.isArray(filenames))
		{
			if (!filenames.length)
			{
				throw new TypeError('Invalid names list. Array must be non-empty.');
			}
		} // names is not an array, nor a string
		else
		{
			throw new TypeError('Invalid type for names! Must be a string or an Array of string.');
		}
	}

	// Deconstruct the source glob into literal and wildcard parts
	let srcGlobParts: IPartsPattern[] = deconstruct(srcGlob, { collapse: true });
	if (!srcGlobParts.length)
	{
		throw new Error('Unexpected error while parsing glob.');
	}

	// Deconstruct the destination glob into literal and wildcard parts
	let dstGlobParts: IPartsPattern[] = deconstruct(dstGlob, { collapse: false });
	if (!dstGlobParts.length)
	{
		throw new Error('Unexpected error while parsing glob.');
	}

	let srcWildcardsCount = countWildcards(srcGlobParts);
	let dstWildcardsCount = countWildcards(dstGlobParts);

	// Check if destination glob has more * and/or ? wildcards than source glob does...
	if ((dstWildcardsCount.stars > srcWildcardsCount.stars) || (dstWildcardsCount.questions > srcWildcardsCount.questions))
	{
		throw new Error('Invalid glob pattern. Destination glob contains incorrect number or type of wildcard. (** is treated as * in source glob.)');
	}

	// extract matches for each wildcard (and literal) parts of the source glob
	let srcCaptureGroupsArray = capture(filenames, srcGlob);
	if (!srcCaptureGroupsArray)
	{
		throw new Error('Unexpected error while extracting glob matches.');
	}

	let computedNames = filenames.map(function buildNewName(name, nameIdx)
	{
		// The source glob is required to match the original name in order to proceed building the new name
		if (!srcCaptureGroupsArray[nameIdx].hasMatch())
		{
			return '';
		}

		let newName = '';
		let srcAsteriskList = srcCaptureGroupsArray[nameIdx].getAsterisk();
		let srcAsteriskIterator = srcAsteriskList?.length ? srcAsteriskList.entries() : null;
		let srcQuestionMarkList = srcCaptureGroupsArray[nameIdx].getQuestionMark();
		let srcQuestionMarkIterator = srcQuestionMarkList?.length
			? srcQuestionMarkList.entries()
			: null;

		// Cycle through the parts of the destination glob to build the new name
		for (const destPart of dstGlobParts)
		{
			let srcGroup;
			switch (destPart)
			{
				case EnumWildcard.Asterisk:
					srcGroup = srcAsteriskIterator.next();
					// value[0] is the key, value[1] is the value, e.g. the actual captureGroup object.
					// Append the * match to the new name construction
					newName = newName + srcGroup.value[1].match;
					break;
				case EnumWildcard.QuestionMark:
					srcGroup = srcQuestionMarkIterator.next();
					// value[0] is the key, value[1] is the value, e.g. the actual captureGroup object.
					// Appends the ? match to the new name construction
					newName = newName + srcGroup.value[1].match;
					break;
				default: // For literal parts, simply append them to new name construction
					newName = newName + destPart;
			}
		}
		return newName;
	});

	return computedNames;
}
