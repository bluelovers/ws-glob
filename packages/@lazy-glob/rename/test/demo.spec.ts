
import { inspect } from 'util';
import { computeName } from '../lib/util/computeName';

describe(`demo`, () =>
{

	([

		['123.456', '*.*', '**'],
		['123.456', '*.*', '*.ts'],
		[[
			'larry-fine-lines-season1-episode3.txt',
			'moe-howard-lines-s04-ep01.txt',
			'curly-howard-lines-s2-e2.txt',
		], '**-*-lines-*?-*?.txt', '*_*-lines-s?e?.txt'],

		[[
			'larry-fine-lines-season1-episode3.txt',
			'moe-howard-lines-s04-ep01.txt',
			'curly-howard-lines-s2-e2.txt',
		], '*.txt', 'temp/*.old'],

	] as Parameters<typeof computeName>[]).forEach(argv => {

		test(inspect(argv), () =>
		{

			let actual = computeName(...argv);

			console.dir(actual);

			expect(actual).toMatchSnapshot();

		});

	})

})
