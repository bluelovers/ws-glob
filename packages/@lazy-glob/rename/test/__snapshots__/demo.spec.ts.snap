// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`demo [
  [
    'larry-fine-lines-season1-episode3.txt',
    'moe-howard-lines-s04-ep01.txt',
    'curly-howard-lines-s2-e2.txt'
  ],
  '**-*-lines-*?-*?.txt',
  '*_*-lines-s?e?.txt'
] 1`] = `
Array [
  "larry_fine-lines-s1e3.txt",
  "moe_howard-lines-s4e1.txt",
  "curly_howard-lines-s2e2.txt",
]
`;

exports[`demo [
  [
    'larry-fine-lines-season1-episode3.txt',
    'moe-howard-lines-s04-ep01.txt',
    'curly-howard-lines-s2-e2.txt'
  ],
  '*.txt',
  'temp/*.old'
] 1`] = `
Array [
  "temp/larry-fine-lines-season1-episode3.old",
  "temp/moe-howard-lines-s04-ep01.old",
  "temp/curly-howard-lines-s2-e2.old",
]
`;

exports[`demo [ '123.456', '*.*', '**' ] 1`] = `
Array [
  "123456",
]
`;

exports[`demo [ '123.456', '*.*', '*.ts' ] 1`] = `
Array [
  "123.ts",
]
`;
